### PR TITLE
Reconcile core data generation features with latest research advances

### PR DESCRIPTION
**Pull request by bbrowning**
This PR brings in some of the latest advancements prototyped by our research team into the broader codebase for everyone's use. It's a work-in-progress, but also something that others may wish to follow, comment on, and contribute to as the work gets done. There are still outstanding features not yet added to this - some new pipeline block types, an improved skills pipeline config, LLMLogProbBlock and LLMMessagesBlock are just stubs, etc.

And, this may cause the deprecation / removal of some existing functionality. That is not entirely clear yet, but will become apparent as work progresses.

See `docs/upgrading_from_v0.6.x_to_v0.7.x.md` for more details, although that too is still only stubbed out.

### CODE REVIEW

**Comment by bbrowning** [comment]
The fact that the e2e-small test passed even though I haven't actually converted any of our default config prompt templates to Jinja syntax yet is concerning, as that means the test is extremely loose in what it considers success. We wouldn't have actually included any of the user's question or answers in the skills or knowledge prompts we sent to the model, and instead it would have all had placeholder python String format tokens in it.

**Comment by bbrowning** [comment]
Ok, I believe this has enough of the research code ported over, new tests created, and existing tests passing that it's ready for review. I know it's a big effort to review this, but it was also a big lift to get the core improvements from research working in our codebase.

BlockRegistry, PromptRegistry, IterBlock, and Jinja prompt templates should all be working. I created #413 and #414 separately to track finishing up `LLMLogProbBlock` and `LLMMessagesBlock`.

This does change some of our public API, although at this time there are no known users of the breaking changes there - basically removal of `ImportBlock` and reorganizing blocks under a `instructlab.sdg.blocks` package. The API used by the InstructLab CLI should be unchanged, and the e2e CI tests still passing without changes confirms this.

**Review by jwm4 — APPROVED** [review]
I don't see any problems with any of these changes, but I am not really an expert (and don't have write access, so my approval doesn't count for much).  I would note that the research code base also has a very nice README file with a lot of useful information.  I would like that merged in too with edits as needed to reflect any differences between that code and the open source code.   However, this PR is already plenty big, so I would recommend a separate PR for the README merge.

**Review by khaledsulayman — COMMENTED** [review]
thanks for this massive lift!

left some minor comments, besides that everything else looks good!

**Comment by bbrowning** [comment]
> I don't see any problems with any of these changes, but I am not really an expert (and don't have write access, so my approval doesn't count for much). I would note that the research code base also has a very nice README file with a lot of useful information. I would like that merged in too with edits as needed to reflect any differences between that code and the open source code. However, this PR is already plenty big, so I would recommend a separate PR for the README merge.

I agree we should pull in those changes - tracked as #428.

File: src/instructlab/sdg/blocks/block.py
Comment (bbrowning) [reply src/instructlab/sdg/blocks/block.py]: Yes, great catch - I just copied the comments over verbatim, but pushed a new commit to adapt these to the style you outlined that we use elsewhere in SDG.

File: src/instructlab/sdg/blocks/block.py
Comment (bbrowning) [reply src/instructlab/sdg/blocks/block.py:45]: You made me realize some of this is leftover from pre-Jinja stuff - we can do away with KeyError and ChainMap entirely here, and just rely on Jinja to throw UndefinedError for us on any undefined keys. Simplifying this logic to do that in another commit.

File: src/instructlab/sdg/blocks/iterblock.py
Comment (aakankshaduggal) [inline src/instructlab/sdg/blocks/iterblock.py:19]: Could we add a docstring to explain what this new block does? It would be helpful to briefly describe its purpose, key attributes like `num_iters`, and how it fits into the overall system. This will make it easier for others to understand its functionality and use. We also don't have to add it in this PR, just a recommendation.

**Review by aakankshaduggal — APPROVED** [review]
Thanks @bbrowning for a great PR! LGTM 🚢

File: src/instructlab/sdg/blocks/iterblock.py
Comment (bbrowning) [reply src/instructlab/sdg/blocks/iterblock.py:19]: Done!

File: src/instructlab/sdg/blocks/llmblock.py
Comment (anastasds) [inline src/instructlab/sdg/blocks/llmblock.py:299]: What's the motivation to use `assert` here instead of raising?

**Review by anastasds — APPROVED** [review]

File: src/instructlab/sdg/blocks/llmblock.py
Comment (bbrowning) [reply src/instructlab/sdg/blocks/llmblock.py:299]: We typically only use `assert` outside of tests to validate/document code assumptions that are not meant to be user-facing or influenced by user config or params. However, in this case, I think you're properly catching that these config_paths are in fact user-facing (as they specify them in their pipeline yaml) and thus these should be a raised exception instead. Will fix :smile:

Comment (bbrowning) [reply src/instructlab/sdg/blocks/llmblock.py]: This now raises a `BlockConfigParserError` instead of asserting, and adds a test to validate that gets raised properly. Thanks for the input!

**Comment by bbrowning** [comment]
Rebased on top of latest main since CI is now fixed there - will let CI chew on things and if it's good and no more comments, going to squash a lot of these into a handful of fewer commits, add the co-authorship metadata for the upstream researchers that did much of the initial code, and get this merged. I'm not planning any functional changes to this PR from this point forward unless more review notes come up.

**Pull request merged**
